### PR TITLE
Create linux-arm64 builds of `kapp-controller`

### LINKED ISSUE #573: Create arm64 image for `kapp-controller`
Currently, `kapp-controller` cannot run on an ARM64 host, as there is no image[0].

Could we switch to doing multi-arch builds and produce arm64 alongside amd64?

[0]:

```
$ crane manifest ghcr.io/vmware-tanzu/carvel-kapp-controller@sha256:a97d9d43b8b5d996f6510ae88ee761ac0f1d883d04fa7402b32f8ce6f9f875bd
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
   "config": {
      "mediaType": "application/vnd.docker.container.image.v1+json",
      "size": 5471,
      "digest": "sha256:57a03345e2ece35c67f3e1ede6cfa10008af841d04a1370e9491475cb31d072b"
   },
   "layers": [
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 17535726,
         "digest": "sha256:259970e9133ac11d271009754f24a11f5f6dd553a4fd7ce214f36a605be71adb"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 102329116,
         "digest": "sha256:38fa5ae806a2cba3ece15743ee641b6ad21c3f2fad4452c1259d759f224d778d"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 177,
         "digest": "sha256:544dd9b3e981ed95d9d45c19ab24ef223d73e1f2302bee286975b000597eed84"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 336772,
         "digest": "sha256:d6c4922f95f2bdf246883dec707a07438ada5bf8ceedd4bc13197abb3648ff02"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 12563601,
         "digest": "sha256:e140a67716edcf8efb267689ef6d9bc800bb30c2b87b7adf8e1802e9b4878f13"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 13737243,
         "digest": "sha256:3c616adb79a7fc3b98ef79a5844290daa8ded835027acd3cf26cd16cfe6bc400"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 10912475,
         "digest": "sha256:b853756a6369238b2b2f94e657b7597f63b0cc917b99c9316582edcea1a4ba3d"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 6437530,
         "digest": "sha256:70ec834fcf4241aea9634bb05119afe861ab51fd4b7a1b2127e26b17205db578"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 7039747,
         "digest": "sha256:82b1d1a8eae1bc1d7a7bce092fd60d4243fb39ae88253b76252f6808f8b50cef"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 8371356,
         "digest": "sha256:4e87932c7ca35afbaa99b438fbdb84ead1f4b62a2a16639c4ed3c041c814b0f1"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 13522494,
         "digest": "sha256:433dd29b575264c4847b163177f6b765db11df38c35a9598360be91f04e031e8"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 2442718,
         "digest": "sha256:95525999d6dbd611b77fabda007fa3129ab8b8a4cbf7e19a5a39c0cf91c34465"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 5510531,
         "digest": "sha256:6fde91471c4ff553a35eee899bed1c01ebac1f9ce309447c0e1ceed5d274b10a"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 21091266,
         "digest": "sha256:3d3bef3596e24879faf108e6d1b3b9b0dfa99445c1483f2882470df490182007"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 29198052,
         "digest": "sha256:ebeb9d7977366aac485ef35389a24f1d9673d1b38ce4097e076d04a28c25e18e"
      }
   ]
}
```

### PR DESCRIPTION
**Pull request by joshrosso**
This PR requires:

* https://github.com/vmware-tanzu/carvel-imgpkg/issues/352
* https://github.com/vmware-tanzu/carvel-vendir/issues/143

to be resolved. For now, I am referencing arm64 binaries of `vendir` and `imgpkg` I have built.

Using this Dockerfile, multi-arch images can be built using:

```
docker buildx build --platform linux/arm64,linux/amd64 --push -t "joshrosso/kapp-controller-multi:v0.x.1" .
```

> change the value of `-t`.

I've pushed a `projects.registry.vmware.com/tce/kapp-controller-multi:v0.30.1` to test. It works for me, on a kind cluster running on arm64.

Resolves: https://github.com/vmware-tanzu/carvel-kapp-controller/issues/573

### CODE REVIEW

File: .github/workflows/test-gh.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ jobs:
       uses: actions/setup-go@v1
       with:
         go-version: "1.17.6"
-    - name: Check out code
+    - name: Check out code into the Go module directory
```
Comment (benmoss) [inline .github/workflows/test-gh.yml]: ?

File: Dockerfile [hunk 2]
```diff
@@ -62,15 +75,18 @@ FROM photon:4.0
 # Install openssh for git
 # TODO(bmo): why do we need sed?
 RUN tdnf install -y git openssh-clients sed
+ARG TARGETARCH
+
+RUN tdnf install -y git openssh-clients shadow-tools sed
```
Comment (benmoss) [inline Dockerfile]: ?

**Review by benmoss — COMMENTED** [review]
Marking as a draft since it's not ready to be merged yet. These ? lines are I think just merge mistakes.

What about using docker buildx for multi-platform images? I wish we could use [ko](https://github.com/google/ko), but our heavy dependency on binaries makes it difficult.

**Comment by joshrosso** [comment]
> Marking as a draft since it's not ready to be merged yet. 


> These ? lines are I think just merge mistakes.

I'll address the merge issues.

> What about using docker buildx for multi-platform images? I wish we could use [ko](https://github.com/google/ko), but our heavy dependency on binaries makes it difficult.

This PR makes changes required to use `docker buildx`:

image

**Comment by benmoss** [comment]
My bad, early morning reading comprehension is not at peak performance

**Comment by joshrosso** [comment]
> My bad, early morning reading comprehension is not at peak performance

All good, I feel your pain.

Let me know if you have suggested approaches for checksum resolution.

These changes make the Dockerfile a bit ugly, i'm happy to add alternative approaches. On the flip-side, this approach is pretty "clear" (ie not clever) in what it's doing IMO.

**Comment by joe-kimmel-vmw** [comment]
i think this will also break our [dev deploy ](https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/hack/dev-deploy.sh#L19) script.   it might even be an invitation to revisit the assumptions made by that script. 

The valuable part of that script is that it builds the executable on the developer's local system which has a ton of cached build artifacts, which saves us a few minutes during that "inner loop" iteration process,  but obviously isn't hermetic the way you'd want for proper testing & releasing.    The rest of that script is just a particular implementation of how to get there.   I can take a look at it later in the week.

**Comment by joshrosso** [comment]
> i think this will also break our [dev deploy ](https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/hack/dev-deploy.sh#L19) script. it might even be an invitation to revisit the assumptions made by that script.
> 
> The valuable part of that script is that it builds the executable on the developer's local system which has a ton of cached build artifacts, which saves us a few minutes during that "inner loop" iteration process, but obviously isn't hermetic the way you'd want for proper testing & releasing. The rest of that script is just a particular implementation of how to get there. I can take a look at it later in the week.

I see that.

I can add architecture detection to that script, as right now it's failing outside of the context of `buildx` since `$TARGETARCH` won't be propagated.

I'm happy to do any work, just want guidance from this team as I don't know how "smart" you want something like this script to become.

**Comment by benmoss** [comment]
- [x] [vendir arm64 build](https://github.com/vmware-tanzu/carvel-vendir/releases/tag/v0.26.0)
- [ ] imgpkg arm64 build (PR has been merged, just waiting on a new release)
- [ ] update release process to release a manifest list

I don't think we need to update `hack/deploy.sh` or `hack/dev-deploy.sh`, we can continue just using single arch  builds, if you set defaults in the `Dockerfile` for `TARGETOS` and `TARGETARCH` it should get the tests passing again. I don't know if kbld supports multi-arch, so some investigation will be needed there since I think that's what we use for releasing right now.

I'm also happy to take this PR over if you want

**Comment by benmoss** [comment]
Closing in favor of #680, apologies for the delay on this

**Pull request closed without merging**